Implement BackgroundService and new Actor class

llucax opened the pull request:
`BackgroundService` is a new abstract base class can be used to write other classes that runs one or more tasks in the background. It provides a consistent API to start and stop these services and also takes care of the handling of the background tasks. It can also work as an `async` context manager, giving the service a deterministic lifetime and guaranteed cleanup.

The new `Actor` class brings quite a few new improvements over the old `@actor` decorator. These are the main differences:

* It doesn't start automatically, `start()` needs to be called to start an actor.
* The method to implement the main logic was renamed from `run()` to `_run()`, as it is not intended to be run externally.
* Actors can have an optional `name` (useful for debugging/logging purposes) and `loop` (if the actor should run in a loop different from the currently running loop).
* The actor will only be restarted if an unhandled `Exception` is raised by `_run()`. It will not be restarted if the `_run()` method finishes normally. If an unhandled `BaseException` is raised instead, it will be re-raised. For normal cancellation the `_run()` method should handle `asyncio.CancelledError` if the cancellation shouldn't be propagated (this is the same as with the decorator).
* The `_stop()` method is public (`stop()`) and will `cancel()` and `await` for the task to finish, catching the `asyncio.CancelledError`.
* The `join()` method is renamed to `wait()`, but they can also be awaited directly ( `await actor`).
* For deterministic cleanup, actors can now be used as `async` context managers.

The base actors (`ConfigManagingActor`, `ComponentMetricsResamplingActor`, `DataSourcingActor`, `PowerDistributingActor`) now inherit from the new `Actor` class, as well as the `MovingWindow`.

Fixes #240, fixes #45, fixes #196.

llucax committed:
Add a `BackgroundService` abstract base class

A `BackgroundService` is a service that runs in the background spawning
one or more tasks. The service can be started and stopped and can work
as an async context manager to provide deterministic cleanup.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add `Actor` abstract base class

This class will replace the `@actor` decorator and it inherits from
`BackgroundService`, providing a consistent interface.

The main differences with the `@actor` decorator are:

* It doesn't start automatically, `start()` needs to be called to start
  an actor.
* The method to implement the main logic was renamed from `run()` to
  `_run()`, as it is not intended to be run externally.
* Actors can have an optional `name` (useful for debugging/logging
  purposes).
* The actor will only be restarted if an unhandled `Exception` is raised
  by `_run()`. It will not be restarted if the `_run()` method finishes
  normally. If an unhandled `BaseException` is raised instead, it will be
  re-raised. For normal cancellation the `_run()` method should handle
  `asyncio.CancelledError` if the cancellation shouldn't be propagated
  (this is the same as with the decorator).
* The `_stop()` method is public (`stop()`) and will `cancel()` and
  `await` for the task to finish, catching the `asyncio.CancelledError`.
* The `join()` is removed an the actor can be awaited directly (like
  `await actor`.
* For deterministic cleanup, actors can now be used as async context
  managers.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Migrate `ConfigManagingActor` to the `Actor` class

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add optional name to actors

This improves actors debuggeability.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
This is a WIP because it is lacking tests and release notes (and updating the rest of the actors). It would be good to have some early feedback while I update the rest of the actors.

Marenz commented on src/frequenz/sdk/actor/_background_service.py:
```diff
@@ -0,0 +1,214 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
```
2023

Marenz commented on src/frequenz/sdk/actor/_background_service.py:
```diff
@@ -0,0 +1,214 @@
+        await self.stop()
+
+    def __await__(self) -> collections.abc.Generator[None, None, None]:
+        """Await this background service.
```
as in, all tasks that still running? Maybe add this info to the docu

llucax replied:
Not sure what's unclear, this implements `await backgroud_service`. I just added a clarification that will wait until all the tasks are finished, hope that makes it more clear.

Marenz commented:
In your commit message, `insted` -> `instead`, `raise` -> `raised` (printed bold in the quoted text)

> The main differences with the `@actor` decorator are:
> [...]
>* The actor will only be restarted if an unhandled `Exception` is **raise**
  by `_run()`. It will not be restarted if the `_run()` method finishes
  normally. If an unhandled `BaseException` is raised **insted**, it will be

Marenz commented on src/frequenz/sdk/actor/_actor.py:
```diff
@@ -0,0 +1,212 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
```
2023

Marenz commented on src/frequenz/sdk/actor/_actor.py:
```diff
@@ -0,0 +1,212 @@
+
+    From [Wikipedia](https://en.wikipedia.org/wiki/Actor_model):
+
+    > An actor as the basic building block of concurrent computation. In response to
```
`as` -> `is`? You quoted it a bit in the middle, here `as` no longer makes sense ;)

llucax replied:
Yeah, good catch. To keep the quote as a quote I changed it to `From Wikipedia an actor is: > [...] the basic building block....`.

llucax commented:
Updated with the suggestions by @Marenz

llucax committed:
Migrate `ComponentMetricsResamplingActor` to the `Actor` class

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Improve `_run()` docs for `ComponentMetricsResamplingActor`

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Make `_ActorInfo` generic

When using an union, if we want to use a method present only in one of
the union types, then we need to assert for the type, so it is better
to just make it generic to the type is always the specific type.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Make `_data_pipeline.initialize()` async

We'll need to start actors in this function in the future, so we need it
to be async to be able to `await actor.start()`.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

Marenz commented:
Other than those small things, design wise I didn't notice any problems with the code. I would leave a more detailed review once the draft part is over.

llucax committed:
Migrate `DataSourcingActor` to the `Actor` class

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Migrate `PowerDistributingActor` to the `Actor` class

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
OK, now this has all actor classes migrated. I'm pretty happy about the results.

Next step are:

- [x] Remove the actor decorator
- [x] Write tests
- [x] Write release notes
- [x] Update PR description

And in a follow-up PR:

- Migrate other classes to use `BackgroundService` (like `MovingWindow`, `BatteryPoolStatus`, and whatever uses some background tasks and has `run`/`stop`/`join` methods.

llucax committed:
Support the `Actor` class in `actor.run()`

When actors deriving from `Actor` are passed to `actor.run()`, they will
be started and awaited to finish too.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Setting the restart limit in `disable_actor_auto_restart` fixture

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Convert actor decorator tests to Actor class tests

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove actor decorator

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Make `MovingWindow` a `BackgroundService`

This gives the moving window a common interface and removes some
bookkeeping from the implementation, as well as making tests more
correct (when using it as an async context manager).

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Reset the actor restart limit in the fixture

Now that actors are stopped properly, we can reestablish the restart
limit after it was changed.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Use a context manager to temporarily change the restart limit

A new utility `actor_restart_limit` context manager is added to be able
to temporarily change the actor's restart limit, and it is now used in
the `disable_actor_auto_restart` fixture.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a new `actor_auto_restart_once` fixture

This is useful to test that actors are being properly restarted.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Improve and add more actor tests

In particular we test that the new restart rules work as expected, and
we also check the logs produced to have extra certainty that what we
expect to happen is really happening.

The restart tests are done using the `run()` function because we expect
to move the restart logic there, so the tests are "forward-compatible".

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Properly propagate `CancelledError` in `wait()`

When a `BackgroundService.wait()` is used, it should also propage
`CancelledError`s, as is it not always normal that a background service
is cancelled unless `stop()` is called, which uses `cancel()` to stop
the tasks.

When `stop()` is used, we properly filter out all `CancelledError`s too,
as in that case we triggered the cancellation, so it is normal.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add an actor cancellation test

We verify that cancelled actors are reported as such by the `run()`
function and that the actor is not automatically restarted.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add `BackgroundService` tests

Add some tests for the `BackgroundService` class. The tests were written
looking at the coverage to avoid duplicating a lot of tests that are
already performed in the `Actor` tests.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add release notes for `BackgroundService` and `Actor`

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Make darglint ignores markdown comments

This is so they don't show up in the generated documentation.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

shsms commented on src/frequenz/sdk/actor/_background_service.py:
```diff
@@ -0,0 +1,307 @@
+                self._resolution_s = resolution_s
+
+            async def start(self) -> None:
+                # Note that you should always use the self._loop attribute access loop
+                # functions instead of asyncio directly.
+                self._tasks.add(self.loop.create_task(self._tick()))
```
the comment says "use self._loop", but the code uses self.loop

llucax replied:
Damn `pylint` didn't catch this... Thanks!

shsms commented on src/frequenz/sdk/actor/_background_service.py:
```diff
@@ -0,0 +1,307 @@
+        self._name: str = str(id(self)) if name is None else name
+        self._loop: asyncio.AbstractEventLoop = (
+            loop if loop else asyncio.get_running_loop()
+        )
```
`loop or asyncio.get_running_loop()` would have been so much cooler.  But I guess the loop parameter is going away for now, until we are clear on why it is needed, as discussed.

Marenz commented:
coonflicts

llucax commented:
> coonflicts

I guess nothing too radical to resolve. I won't update the PR until the review is done, as it is a big PR and it could be very annoying to have updates while reviewing.

shsms commented on tests/actor/test_resampling.py:
```diff
@@ -122,39 +122,37 @@ async def test_single_request(
+            data_source_chan_name=data_source_req.get_channel_name(),
+        )
+
+        await resampling_actor._resampler.stop()  # pylint: disable=protected-access
```
probably not relevant to the PR, but just to understand, is it not possible to make the resampler stop, when the actor stops as well?

Maybe we need a cleanup method in the `BackgroundService` or `Actor` base class that gets called before the `stop` of the background service,  that actors can override if they have additional cleanups?  Or maybe the actor is able to get the background service to manage the task of the resampler as well.

llucax replied:
Yes, we can override the `stop()` method itself, I'm not sure why I didn't do it for the resampler, I think I did it for some other actor or background service (probably the moving window).

I will update!

llucax replied:
Not so simple: This is related to the comment in the resampling actor's `_run()` cleanup (`finally`):

```py
            # XXX: Here we should probably do a:  pylint: disable=fixme
            # await self._resampler.stop()
            # But since the actor will be restarted, the internal state would
            # be broken if we stop the resampler.
            #
            # We have an even bigger problem with this naive restarting
            # approach, as restarting this actor without really resetting its
            # state would be mostly the same as not really leaving the run()
            # method and just swallow any exception, which doesn't look super
            # smart.
```

This is why I'm planning to move the restart logic from the `Actor` to the `run()` method. Since the actor is monitoring himself, there is one task that is special, as we can't kill it because is the task monitoring the `_run()` method and restarting it. The self-auto-restarting makes everything more complicated and it is not very resilient,as the actor's internal state is not reset, so if it was left in an inconsistent state because of some bug, re-starting the `_run()` won't fix it. Actually restarting it in `run()` fixes part of the problem but not this, maybe we need some explicit `restart()` method that cleans up the internal state.

So I would leave this for future PRs, I also didn't converted more classes to background services (the resampler and its helper classes should be also background services) because they are not breaking changes, and I wanted to prioritize breaking changes for 1.0.

shsms commented on src/frequenz/sdk/actor/_background_service.py:
```diff
@@ -0,0 +1,307 @@
+            loop: The event loop this background service will run on. If `None`, the
+                current running event loop will be used.
+        """
+        self._name: str = str(id(self)) if name is None else name
+        self._loop: asyncio.AbstractEventLoop = (
+            loop if loop else asyncio.get_running_loop()
+        )
+        self._tasks: set[asyncio.Task[Any]] = set()
```
I wonder if `self._name` and `self._tasks` can be **also** initialized in the `start` method,  so that people don't have to call `super.__init__()` from actor init methods,  even if they don't want to set a name.

(oops, wanted to add to review, but clicked wrong button, because reviewing in multiple tabs)

llucax replied:
I don't understand this. You mean initialize it if it wasn't set in the constructor?

In any case I think it will make things more complicated, for example the `run()` function emits some logs that actors will be started before they are started, so in that case they won't have the names set. Also we'll need to make `_name` and `_tasks` optional (`type | None`, which will force us to litter the implementation with `assert _ not None`). I think having to write the `super().__init__()` is not that bad, also some of the linters (I don't remember if it is `mypy` or `pylint`) catches the issue it if you forget.

shsms replied:
> You mean initialize it if it wasn't set in the constructor?

yup,  but if pylint, etc is able to catch a missing `super().__init__()` that is just as good.  Is there a test for this case already?

llucax replied:
You mean a test that `pylint` catches it?

shsms replied:
yup

shsms replied:
Oh,  but pylint doesn't complain if there are unused `disable=` in the code.  Unlike mypy which complains if there are unused `type: ignore`.  So if the feature is provided by pylint, then maybe it is hard to test.

llucax replied:
Yeah, I'm not sure how I would be able to test if pylint keeps catching this 😕

llucax replied:
BTW, it is pylint that catches this: https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/super-init-not-called.html

shsms commented on src/frequenz/sdk/actor/_background_service.py:
```diff
@@ -0,0 +1,307 @@
+            loop: The event loop this background service will run on. If `None`, the
+                current running event loop will be used.
+        """
+        self._name: str = str(id(self)) if name is None else name
```
can we get it to say have the name of the class of `self`, followed by the id,  to make it a bit more readable?

llucax replied:
This is what `str` (and `repr`) do. I think having the class name always in `str` and `repr` is a good thing, so it will be duplicate if I add it to the name too.

shsms commented on benchmarks/timeseries/periodic_feature_extractor.py:
```diff
@@ -27,19 +29,23 @@
+    period: int,
+) -> collections.abc.AsyncIterator[PeriodicFeatureExtractor]:
     """Initialize the PeriodicFeatureExtractor class."""
     # We only need the moving window to initialize the PeriodicFeatureExtractor class.
     lm_chan = Broadcast[Sample[Quantity]]("lm_net_power")
-    moving_window = MovingWindow(
+    async with MovingWindow(
         timedelta(seconds=1), lm_chan.new_receiver(), timedelta(seconds=1)
-    )
-
-    await lm_chan.new_sender().send(Sample(datetime.now(tz=timezone.utc), Quantity(0)))
+    ) as moving_window:
+        await lm_chan.new_sender().send(
+            Sample(datetime.now(tz=timezone.utc), Quantity(0))
+        )
 
-    # Initialize the PeriodicFeatureExtractor class with a period of period seconds.
-    # This works since the sampling period is set to 1 second.
-    return PeriodicFeatureExtractor(moving_window, timedelta(seconds=period))
+        # Initialize the PeriodicFeatureExtractor class with a period of period seconds.
+        # This works since the sampling period is set to 1 second.
+        yield PeriodicFeatureExtractor(moving_window, timedelta(seconds=period))
```
looks like this usage of `asynccontextmanager` will mean this function can't be used for creating two simultaneous feature extractor instances, for example,  because when a second call is made,  the first one and its corresponding `MovingWindow` will get cleaned up?

So it has to used only with a `async with` right?

Because I can easily imagine people wanting to create multiple PeriodicFeatureExtractor instances for different data streams, etc. because they might want multiple features.

shsms replied:
I wonder if it is possible to just make PeriodicFeatureExtractor tie its lifetime with that of its corresponding moving window.  But I guess for that it would need full ownership of the MovingWindow,  which we are not giving.  

People might want to keep using the same MovingWindow instance after or independent of the life time of the PeriodicFeatureExtractor.

llucax replied:
Not sure if I'm not understanding the issue here but please note that this is just the benchmark, I used `asynccontextmanager` because it was convenient in this context, but users don't have to do the same. The periodic feature extractor itself is not a context manager.

shsms replied:
it was just about what impact the use of `asynccontextmanager` would have if the decorated function has an `async with` inside it, when the decorated function is called multiple times.

Anyway, just wanted to make sure we don't give people the wrong idea about how this can be used,  if people try to read the benchmark script and try to take some inspiration from it.

But it can be examined later and resolved if it is a problem.

llucax replied:
Oh, ok, if the context manager is used more than once, then it is executed twice and will create a new periodic feature extractor. I think everything should work as expected

shsms reviewed (commented):
big PR, but still it was very easy to review because of the well organized commits!

llucax commented:
> big PR, but still it was very easy to review because of the well organized commits!

Yeah, I think it was a **huge** win in terms of reviewing dropping the requirement of every commit being a working piece and focusing more on splitting commits to make changes very isolated. Very glad we tried it out!

llucax committed:
Improve release notes formatting

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Updated:

* Rebased on the current head (I force pushed the rebase separately, so it is easier to see the updates separately from the [rebase diff](https://github.com/frequenz-floss/frequenz-sdk-python/compare/f0bc7ba9a311778c5a985fa67cf490e119183514..59eb30c838bafae769acb093ba8be90c37adaaa2)).
* Removed the `loop` argument from both `BackgroundService` and `Actor`
* Removed the *complete* example from `BackgroundService` and the simple example now forwards the `name` argument
* Same with the README, now only one example is provided for the migration.
* Merged the commit [Add names to MovingWindow tasks](https://github.com/frequenz-floss/frequenz-sdk-python/pull/564/commits/ebebdccc92b092a6f183119a4f6380314b301c32) into [Make MovingWindow a BackgroundService](https://github.com/frequenz-floss/frequenz-sdk-python/pull/564/commits/98eecef28ca8b436553e38201a3d0492d31f93b9)
* Squashed `fixup!` commits
* Other minor fixes
* [Changes diff](https://github.com/frequenz-floss/frequenz-sdk-python/compare/59eb30c838bafae769acb093ba8be90c37adaaa2..ace41e1f9dd4d904e17c80e0843039740d8c36c4)

llucax commented:
Enabling auto-merge

shsms reviewed (approved):
Awesome, everybody get ready to refactor everything now!

shsms commented:
> Enabling auto-merge

I think you also have to click a button,  looks like you just typed the message. :D

llucax commented:
It seems like GitHub copilot is not there yet...
